### ISSUE TITLE
Incomplete repr/str for polynomial models

### ISSUE THREAD

**Issue by jehturner** [astropy/astropy#9795]
**Not a bug -- see below.**

This happens with at least Chebyshev1D/2D in the latest `master` branch `54f9e54d`). I haven't tried other models yet. The resulting model reports incorrect parameters *but* gets evaluated as if the correct parameters are being used (am I doing anything unusually stupid?):

```python
>>> import numpy as np
>>> from astropy.modeling import models, fitting

>>> x = np.arange(0, 10)

>>> fitter = fitting.LinearLSQFitter()

>>> c1 = models.Chebyshev1D(1, c0=0.1, c1=0.5)

>>> # This should reproduce c1 within rounding errors, but appears not to:
>>> c2 = fitter(models.Chebyshev1D(1), x, c1(x))
>>> print('C2', repr(c2))
C2 <Chebyshev1D(1, c0=2.35, c1=2.25)>

>>> # This has the same parameter values reported by c2 (so should do the same):
>>> c3 = models.Chebyshev1D(1, c0=2.35, c1=2.25)
>>> print('C3', repr(c3))
C3 <Chebyshev1D(1, c0=2.35, c1=2.25)>

>>> print('C1(1)', c1(1))
C1(1) 0.6
>>> print('C2(1)', c2(1))  # behaves like c1, not according to its parameters
C2(1) 0.5999999999999972
>>> print('C3(1)', c3(1))  # behaves like c2 would be expected to
C3(1) 4.6
```

Linux-3.10.0-957.27.2.el7.x86_64-x86_64-with-centos-7.6.1810-Core
Python 3.6.9 |Anaconda, Inc.| (default, Jul 30 2019, 19:07:31) 
[GCC 7.3.0]
Numpy 1.17.3
Scipy 1.3.1

**Comment by jehturner** [comment]
@nden, @perrygreenfield

**Comment by jehturner** [comment]
I see the same behaviour when I `git checkout v4.0rc1`, @bsipocz. But to my surprise I'm also seeing it when I check out 2.0 & 3.0. Am I imagining things?

**Comment by jehturner** [comment]
OK, this is not a bug, there's just a difference in `domain` for `c1`/`c3` (which default to `[0,1]`) and `c2` (which gets a domain spanning the input co-ordinates). Sorry for the false alarm, but I think before closing this we should check whether this is documented clearly enough, since there are probably other users who are just as slow as I am. I'm not seeing much in the docs, but I'm not as familiar with the reorganized version.

**Comment by jehturner** [comment]
The docs list the domain parameter(s) for the relevant models, but seem to offer almost no further explanation (which could include how `domain` is related to `window`). However, we're thinking the main issue here is that the domain should be included in the `__repr__()` and `__str__()` output, otherwise those representations are incomplete and do not specify the (polynomial) model uniquely, which is misleading. Conversely, including that information would make this issue obvious without relying on users reading the docs from cover to cover... It looks like this can easily be accomplished by adding some keyword args to the `_format_repr()` & `_format_str()` calls in `OrthoPolynomialBase` and `PolynomialModel`. How does that sound?

**Comment by jehturner** [comment]
Oh, but `PolynomialModel` knows nothing about `domain` (unlike `OrthoPolynomialBase`). Would it be acceptable for the former to pass the `domain` value to `_format_repr()` when `hasattr(self, 'domain')`, or should we override `__repr__()` and `__str__()` separately in every 1D polynomial model because that's where the `domain` argument is introduced?

**Comment by jehturner** [comment]
Sounds good, though I notice that `LinearLSQFitter()` does use `hasattr(model, 'domain')`, so there's already some expectation of that information being used outside the scope of the relevant models and it would therefore seem fairly harmless in `PolynomialModel`.

Some of the model `__init__()`s might disappear, since `domain` (& `window`) would be introduced in the new 1D base class, though at least some of them will still need to overload `__init__()` due to differing default values.

I'm not sure I'll get back to this before the New Year now, but I don't think there's a special hurry, is there? Thanks.

**Comment by nden** [comment]
Using `hasattr` should be fine too. And no, there's no rush. Thanks for working on this!

**Comment by jehturner** [comment]
This gets worse, @nden. It appears that the asdf model serialization is not including the domain information either :slightly_frowning_face:. I'm just looking into it a bit more. That should probably have its own issue, though the 2 things are somewhat related.

**Comment by nden** [comment]
Right. Too bad we didn't think about it before the 4.0 release. It should go in the first patch release. Please add an issue about it.

**Comment by nden** [comment]
Ah, I see you did

**Comment by jehturner** [comment]
Yeah, sorry, I posted it by mistake before adding the details.

**Comment by jluastro** [comment]
Just wanted to second the suggestion above that there should be more documentation on domain and window and their usage (in Polynomial2D for instance).

**Comment by eteq** [comment]
I'm a bit lost in the intertwining of multiple issues here.  I think there are three distinct issues here, one of which has a PR, but the others of which are still pending :

* The ASDF serialization issue (#9809)
* Documentation clarifying exactly when window vs domain is used (👍'ed by  @jluastro's comment above) 
* Having the domain/window appear in the `repr`/`str` so that a polynomial is replicable from the string representation (which I think is a reasonable expectation).

I think those latter two are nominally separate issues (the third one I'm particularly focused on right now because it just bit me!).  Should I create follow-ons for those, or did I miss something (@nden  @jehturner @perrygreenfield)?

**Comment by jehturner** [comment]
Yes, that sounds right, @eteq. I was just getting back to looking at the `__repr__` as it happens.

**Comment by perrygreenfield** [comment]
I'm trying to sort all this out now. There are some cases where the behavior is perhaps not what one would expect. For example, the default definitions for both domain and window for general polynomials is [-1, 1]. One might conclude that only values in that range are allowed, but that isn't the case. All it means is that there is no transformation of the input value (e.g., if the domain were [-2, 2] and window was [-1, 1] then the input variable would be divided by 2 before being used in the polynomial). In that particular case, putting it in the repr and str is more misleading than helpful. But in any case, I'd agree that it may not be what users expect (bounding_box is now used to exclude values input values. Perhaps a different formalism would be more appropriate, such as input_transform that specifies the linear coefficients. But then that entails a transition from one scheme to another.

**Comment by perrygreenfield** [comment]
@eteq  I think the first and last item are inextricably linked in that they need to be consistent with each other.

**Comment by jehturner** [comment]
Gah, yes, so documentation would be really useful. It wasn't clear to me what exactly the window does. Do you want to take care of the `repr` as well then? I was about to fix that, but you're the expert and I'm still not sure I understand all the details... Why would `domain` & `window` in the repr be more misleading than helpful in your example above?

**Comment by perrygreenfield** [comment]
Because some might read domain as the range of legitimate input values, but in that case, it doesn't mean that at all. It is only used to define the transform to be applied to input values (i.e., the linear transform that maps the domain range into the window range).

**Comment by perrygreenfield** [comment]
In this it would be good to get feedback on what users think is the best way to define these things. At the moment for orthonormal polynomials, window is used to indicate that the allowable range of input values is  [-1, 1], but I don't think that is actually enforced (but I will have to check). So what we have is some overloading of these values for different purposes, which I also think may lead to some confusion.

**Comment by jehturner** [comment]
OK (to the last but one comment). I wondered if that's what you meant, but isn't that more a naming or documentation issue than a `repr` issue? I would second @eteq's reasonable expectation that the string representation should unambiguously define the behaviour as far as practical (at least when it *looks* like it does).

**Comment by jehturner** [comment]
Yes, I think `window` is not enforced, having played with it a bit to try to understand what it does.

**Comment by perrygreenfield** [comment]
Oh, absolutely, the string representation should be unambiguous (and correct too!), and in my view clear as well.

**Comment by perrygreenfield** [comment]
Nadia points out that this approach (domain and window) is what numpy uses for its orthonormal polynomials, and it does permit values outside of the window, so there is a consistency argument there. (Though I do wonder about permitting x values outside the window for polynomials specifically designed for a range of [-1, 1].)

**Comment by jehturner** [comment]
That seems reasonable (and favours documentation), but I'm not sure I see the point of `window` if it's not enforced(?).

What you say about x values outside the defined range makes sense, but I imagine it's going to be common to fit a curve to data that don't quite span some pixel range and then need to evaluate the fit slightly out of range at the ends? Disallowing that might cause more trouble than it solves. Or am I doing it wrong?

**Comment by jehturner** [comment]
Looks like it's not just us that are confused:
https://github.com/numpy/numpy/issues/9533

**Comment by perrygreenfield** [comment]
Funny you should mention that since one thing that happens in fitting is to define the domain to be the min max of the input points. For example, see line 288 in astropy/modeling/fitting.py

There if the domain is None, the fitter defines the domain to match the min max of the points to be fitted.

**Comment by perrygreenfield** [comment]
Regarding numpy/numpy#9533 I have to say I am not surprised. I was going to look myself, but you saved me the trouble.

**Comment by jehturner** [comment]
Note Chuck's comment on `window` in https://github.com/numpy/numpy/issues/9533#issuecomment-321595955.

**Comment by perrygreenfield** [comment]
In other words, we made a boo-boo. I think we would all agree that separating the transform definition from domain and window would have been a good thing. Question is, should we now attempt to do that?

**Comment by perrygreenfield** [comment]
One approach might be to see if the transform terms are defined, in which case that is used in place of the domain and window definitions, which then drive different interpretations of window and domain. That complicates the code, but may provide a path to a more sensible interface.

**Comment by perrygreenfield** [comment]
To expand on this approach a bit more, we could define a new attribute for polynomials input_transform (x_,y_input_transform for 2d) which is a tuple that the x_new = input_transform[1] * x_old + input_transform[0]

if the value of that attribute is None it implies [0, 1]. We could use None to as a signal to use the old scheme, but it would be nice eventually for it to mean no transform. I guess that implies another attribute that indicates which
scheme to use. Defining input_transform would set that attribute to True, but one could also use a keyword argument to set that attribute on initialization.

For the new scheme, domain is not used at all (bounding_box serves that purpose now, in that it defines what range of input values are allowed). Window set to None indicates that all transformed input values are permitted. One would also be able to indicate open ends, e.g., window=(0, None) means all values 0 and greater are permitted. There would be another attribute (and corresponding keyword argument) that says one can ignore window on evaluation. For example, in using orthonormal polynomials, the default window is (-1, 1) and when supplied any value outside that range, an exception results. But one can provide a keyword argument to the call to say ignore the window, or in the definition of the orthonormal polynomial that window can be ignored. That way people normally would never modify the default value of window for these polynomials (We could even prohibit changing them), but allow flexibility in whether they are enforced.

In the absence of a the new_scheme attribute=True, the old behavior continues to work (perhaps deprecated in the future, but not necessarily). The documentation should strongly discourage its use as confusing, though.

**Comment by jluastro** [comment]
How long has the window/domain version been around? Personally, I found it too confusing to use much and did all the data rescaling first to the -1 1 range before putting into models. Thus I am wondering how much we would actually break if people haven’t been using these features much anyhow. However, if numpy was doing something similar, then there might have been people that found there way to that documentation. 

For real-world use, we would definitely like the capability to fit on a data set (with min/max data range rescaled to +/-1 for well-behaved polynomial fitting) and then apply to a data set that is slightly outside that range. This is very common. I agree that it would be nice to have this optional or some kind of warning if the model is being extrapolated.

**Comment by jehturner** [comment]
I'll need to think about this, but it seems there are 2 timescales to think about: bug fixes for `4.0.1` (including #9906) and any API changes for >= 4.1 (scheduled for April)? The `repr` changes could probably go in either, as long as we're clear what we want it to look like, which I think is what you're trying to figure out now.

**Comment by bsipocz** [comment]
> The repr changes could probably go in either, as long as we're clear what we want it to look like, which I think is what you're trying to figure out now.

I would think the repr changes are also 4.1 material in case the current setup is not considered utterly broken, projects might rely on it in doctests, etc that we shouldn't break in a bugfix release.

**Comment by nden** [comment]
I agree with @jehturner that we need to fix the serialization for 4.0.1 . We should also clarify `domain` and `window` in the documentation. API changes can go in 4.1 if agreed on. Note that the numpy issue was created in 2017 and is still open.
